janus_server/src/trace.rs

```diff
@@ -1,40 +1,77 @@
+use atty::{self, Stream};
+use tracing_log::LogTracer;
 use tracing_subscriber::{fmt, layer::SubscriberExt, EnvFilter, Registry};
 
+/// Errors
 #[derive(Debug, thiserror::Error)]
 pub enum Error {
     #[error("tracing error: {0}")]
-    HttpClient(#[from] tracing::subscriber::SetGlobalDefaultError),
+    SetGlobalTracingSubscriber(#[from] tracing::subscriber::SetGlobalDefaultError),
+    #[error("logging error: {0}")]
+    SetGlobalLogger(#[from] tracing_log::log_tracer::SetLoggerError),
+}
+
+/// Configuration for the tracing subscriber
+#[derive(Debug, Default)]
+pub struct TraceConfiguration {
+    /// If true, uses a [`tracing_subscriber::fmt::TestWriter`] to capture trace
+    /// events when running tests
+    pub use_test_writer: bool,
+    /// If true OR if stdout is a tty, trace events are output in JSON format by
+    /// [`tracing_subscriber::fmt::format::Json`]. Otherwise, trace events are
+    /// output in pretty format by [`tracing_subscriber::fmt::format::Pretty`].
+    pub force_json_output: bool,
 }
 
 /// Configures and installs a tracing subscriber
-pub fn install_subscriber() -> Result<(), Error> {
+pub fn install_subscriber(config: &TraceConfiguration) -> Result<(), Error> {
     // Configure a tracing subscriber. The crate emits events using `info!`,
     // `err!`, etc. macros from crate `tracing`.
     let fmt_layer = fmt::layer()
         .with_thread_ids(true)
-        // TODO(#16): take an argument for pretty vs. full vs. compact vs. JSON
-        // output
-        .pretty()
         .with_level(true)
-        .with_target(true);
+        .with_target(true)
+        .pretty();
+
+    if config.use_test_writer {
+        fmt_layer = fmt_layer.with_test_writer();
+    }
+
+    // If stderr is not a tty or if forced by config, output logs as JSON
+    // structures
+    if atty::isnt(Stream::Stdout) || config.force_json_output {
+        fmt_layer = fmt_layer.json();
+    };
```

Comment:
I am having trouble getting this part to work and could use some help. You can't do this because `fmt_layer` and `fmt_layer.with_test_writer()` don't have the same type. Normally I would deal with this by boxing the values ([as I did with `slog::Drain`s in `prio-server`](https://github.com/divviup/prio-server/blob/9faa6e2566bb9c5b590e85d223f08e34b8e1d95a/facilitator/src/logging.rs#L130)) but that doesn't work here.

For one thing, even if I could box these into `Box<dyn Layer<...>>`, [`trait Layer`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/trait.Layer.html#) doesn't have the same methods as [`struct Layer`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/struct.Layer.html#) that would let me chain on a call to `.json()`. 

I must be missing something about how to compose layers together. I also tried creating multiple layers and layering them together into [`Layered`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/struct.Layered.html), but the result is that you get events going to multiple places -- stdout with `pretty` and then also stdout with `json`.